x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/ResultsIndexUpgradeAction.java

```
@@ -0,0 +1,163 @@
+        static {
+            PARSER.declareInt(Request::setReindexBatchSize, REINDEX_BATCH_SIZE);
+        }
+
+        static final String INDEX = AnomalyDetectorsIndexFields.RESULTS_INDEX_PREFIX + "*";
+        private int reindexBatchSize = 1000;
+
+        /**
+         * Should this task store its result?
+         */
+        private boolean shouldStoreResult;
+
+        // for serialization
+        public Request() {
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            reindexBatchSize = in.readInt();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeInt(reindexBatchSize);
+        }
+
+        @Override
+        public String[] indices() {
+            return new String[]{INDEX};
+        }
+
+        @Override
+        public IndicesOptions indicesOptions() {
+            //TODO consider lenientExpandOpen() ?
+            return IndicesOptions.strictExpandOpenAndForbidClosed();
+        }
+
+        /**
+         * Should this task store its result after it has finished?
+         */
+        public ResultsIndexUpgradeAction.Request setShouldStoreResult(boolean shouldStoreResult) {
+            this.shouldStoreResult = shouldStoreResult;
+            return this;
+        }
+
+        @Override
+        public boolean getShouldStoreResult() {
+            return shouldStoreResult;
+        }
+
+        public ResultsIndexUpgradeAction.Request setReindexBatchSize(int reindexBatchSize) {
+            this.reindexBatchSize = reindexBatchSize;
+            return this;
+        }
+
+        public int getReindexBatchSize() {
+            return reindexBatchSize;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+           return null;
```

Maybe validate that `reindexBatchSize` is >= 0?  The search request created by the reindex will throw an exception if its size is negative, but by the time that happens the request will have been transported to the node it's going to run on, and it's best to fail fast.